tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -12,889 +13,945 @@
+        {
+            ThrowHelper.ThrowArgumentException($"The method: {BeginMethodName} with {onMethodBeginParameters.Length} parameters in type: {integrationType.FullName} has less parameters than required.");
+        }
+        else if (onMethodBeginParameters.Length > argumentsTypes.Length + 1)
+        {
+            ThrowHelper.ThrowArgumentException($"The method: {BeginMethodName} with {onMethodBeginParameters.Length} parameters in type: {integrationType.FullName} has more parameters than required.");
+        }
+        else if (onMethodBeginParameters.Length != argumentsTypes.Length && onMethodBeginParameters[0].ParameterType != genericArgumentsTypes[0])
+        {
+            ThrowHelper.ThrowArgumentException($"The first generic argument for method: {BeginMethodName} in type: {integrationType.FullName} must be the same as the first parameter for the instance value.");
+        }
+
+        var callGenericTypes = new List<Type>();
+        var mustLoadInstance = onMethodBeginParameters.Length != argumentsTypes.Length;
+        var instanceGenericType = genericArgumentsTypes[0];
+        var instanceGenericConstraint = instanceGenericType.GetGenericParameterConstraints().FirstOrDefault();
+        Type? instanceProxyType = null;
+        if (instanceGenericConstraint != null)
+        {
+            var result = DuckType.GetOrCreateProxyType(instanceGenericConstraint, targetType);
+            instanceProxyType = result.ProxyType;
+            if (instanceProxyType is null)
             {
-                ThrowHelper.ThrowArgumentException($"The first generic argument for method: {BeginMethodName} in type: {integrationType.FullName} must be the same as the first parameter for the instance value.");
+                ThrowHelper.ThrowArgumentException($"The instance proxy type for method: {BeginMethodName} in type: {integrationType.FullName} is null.");
             }
 
-            List<Type> callGenericTypes = new List<Type>();
+            callGenericTypes.Add(instanceProxyType);
+        }
+        else
+        {
+            callGenericTypes.Add(targetType);
+        }
 
-            bool mustLoadInstance = onMethodBeginParameters.Length != argumentsTypes.Length;
-            Type instanceGenericType = genericArgumentsTypes[0];
-            Type instanceGenericConstraint = instanceGenericType.GetGenericParameterConstraints().FirstOrDefault();
-            Type instanceProxyType = null;
-            if (instanceGenericConstraint != null)
-            {
-                var result = DuckType.GetOrCreateProxyType(instanceGenericConstraint, targetType);
-                instanceProxyType = result.ProxyType;
-                callGenericTypes.Add(instanceProxyType);
-            }
-            else
-            {
-                callGenericTypes.Add(targetType);
-            }
+        var callMethod = new DynamicMethod(
+            $"{onMethodBeginMethodInfo.DeclaringType!.Name}.{onMethodBeginMethodInfo.Name}",
```

Comment:
_sigh_ I hate all the DAMMIT! operators you need with reflection types :(

Reply:
:nod_cry: